packages/expo-updates/ios/EXUpdates/CodeSigning/EXUpdatesCertificateChain.swift

```diff
@@ -0,0 +1,194 @@
+  
+  public required init(certificateStrings: [String]) throws {
+    self.certificateStrings = certificateStrings
+  }
+  
+  public func codeSigningCertificate() throws -> Certificate {
+    if (certificateStrings.isEmpty) {
+      throw EXUpdatesCodeSigningError.CertificateEmptyError
+    }
+    
+    let certificateChain = try certificateStrings.map { certificateString throws in
+      try EXUpdatesCertificateChain.constructCertificate(certificateString: certificateString)
+    }
+    try certificateChain.validateChain()
+    
+    let leafCertificate = certificateChain.first!
+    let (_, x509lLeafCertificate) = leafCertificate
+    if (!x509lLeafCertificate.isCodeSigningCertificate()) {
+      throw EXUpdatesCodeSigningError.CertificateMissingCodeSigningError
+    }
+    
+    return leafCertificate
+  }
+  
+  private static func constructCertificate(certificateString: String) throws -> Certificate {
+    guard let certificateData = certificateString.data(using: .utf8) else { throw EXUpdatesCodeSigningError.CertificateParseError }
+    
+    guard let certificateDataDer = decodeToDER(pem: certificateData) else {
+      throw EXUpdatesCodeSigningError.CertificateParseError
+    }
+    
+    let x509Certificate = try X509Certificate(der: certificateDataDer)
+    
+    guard x509Certificate.checkValidity() else {
+      throw EXUpdatesCodeSigningError.CertificateValidityError
+    }
+    
+    guard let secCertificate = SecCertificateCreateWithData(nil, certificateDataDer as CFData) else {
+      throw EXUpdatesCodeSigningError.CertificateParseError
+    }
+    
+    return (secCertificate, x509Certificate)
+  }
+  
+  private static let beginPemBlock = "-----BEGIN CERTIFICATE-----"
+  private static let endPemBlock   = "-----END CERTIFICATE-----"
+  
+  /**
+   * Mostly from ASN1Decoder with the fix for disallowing multiple certificatess in the PEM.
+   */
+  private static func decodeToDER(pem pemData: Data) -> Data? {
+    guard let pem = String(data: pemData, encoding: .ascii) else {
+      return nil
+    }
+    
+    if pem.components(separatedBy: beginPemBlock).count - 1 != 1 {
+      return nil
+    }
+    
+    let lines = pem.components(separatedBy: .newlines)
+    var base64buffer  = ""
+    var certLine = false
+    for line in lines {
+      if line == endPemBlock {
+        certLine = false
+      }
+      if certLine {
+        base64buffer.append(line)
+      }
+      if line == beginPemBlock {
+        certLine = true
+      }
+    }
+    if let derDataDecoded = Data(base64Encoded: base64buffer) {
+      return derDataDecoded
+    }
+    
+    return nil
+  }
+}
+
+extension X509Certificate {
+  func isCACertificate() -> Bool {
+    if let ext = self.extensionObject(oid: .basicConstraints) as? X509Certificate.BasicConstraintExtension {
+      if (!ext.isCA) {
+        return false
+      }
+    } else {
+      return false
+    }
+    
+    let keyUsage = self.keyUsage
+    if (keyUsage.isEmpty || !keyUsage[5]) {
```

Comment:
Can `keyUsage[5]` trigger a bounds check error?

Reply:
looking at the code, this is guaranteed to have 7 indices. Also this code is the same as android.